Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt

```diff
@@ -0,0 +1,79 @@
+/*
+ * Paintroid: An image manipulation application for Android.
+ * Copyright (C) 2010-2021 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.catrobat.paintroid.tools.helper
+
+import android.graphics.Bitmap
+import android.graphics.Color
+import android.graphics.Rect
+import android.util.Log
+
+private const val TAG = "cropAlgorithmSnail"
+
+class JavaCropAlgorithm : CropAlgorithm {
+    private lateinit var bounds: Rect
+    private lateinit var bitmap: Bitmap
+
+    private fun Rect.rowRange() = this.top..this.bottom
+
+    private fun Rect.colRange() = this.left..this.right
+
+    private fun isOpaquePixel(x: Int, y: Int): Boolean = bitmap.getPixel(x, y) != Color.TRANSPARENT
+
+    private fun getTopmostCellWithOpaquePixel(): Int? =
+        bounds.rowRange().firstOrNull { y ->
+            bounds.colRange()
+                .firstOrNull { x -> isOpaquePixel(x, y) } != null
+        }
+
+    private fun getBottommostCellWithOpaquePixel(): Int? =
+        bounds.rowRange().lastOrNull { y ->
+            bounds.colRange()
+                .firstOrNull { x -> isOpaquePixel(x, y) } != null
+        }
+
+    private fun getLeftmostCellWithOpaquePixel(): Int? =
+        bounds.colRange().firstOrNull { x ->
+            bounds.rowRange()
+                .firstOrNull { y -> isOpaquePixel(x, y) } != null
+        }
+
+    private fun getRightmostCellWithOpaquePixel(): Int? =
+        bounds.colRange().lastOrNull { x ->
+            bounds.rowRange()
+                .firstOrNull { y -> isOpaquePixel(x, y) } != null
+        }
+
+    override fun crop(bitmap: Bitmap?): Rect? {
+        if (bitmap == null) {
+            Log.e(TAG, "bitmap is null!")
+            return null
+        }
+
+        this.bitmap = bitmap
+        bounds = Rect(0, 0, bitmap.width - 1, bitmap.height - 1)
+
+        val top = getTopmostCellWithOpaquePixel() ?: return null
+        bounds.top = top
+        bounds.bottom = getBottommostCellWithOpaquePixel() ?: bounds.top - 1
+        bounds.left = getLeftmostCellWithOpaquePixel() ?: bounds.right + 1
+        bounds.right = getRightmostCellWithOpaquePixel() ?: bounds.left - 1
```

Comment:
The right side of each ?: won't be executed, and if there would be a bug in the algorithm.
Anyway, I think it's ok :+1: